packages/SwingSet/src/message.js

```diff
@@ -28,4 +28,160 @@ export function insistMessage(message) {
       X`message has non-string non-null .result ${message.result}`,
     );
   }
+  return undefined;
+}
+
+/**
+ * @param {unknown} vdo
+ * @returns { VatDeliveryObject, asserts vdo is VatDeliveryObject }
+ */
+export function insistVatDeliveryObject(vdo) {
+  assert(Array.isArray(vdo));
+  const [type, ...rest] = vdo;
+  switch (type) {
+    case 'message': {
+      const [target, msg] = rest;
+      assert.typeof(target, 'string');
+      insistMessage(msg);
+      break;
+    }
+    case 'notify': {
+      const [resolutions] = rest;
+      assert(Array.isArray(resolutions));
+      for (const [vpid, rejected, data] of resolutions) {
+        assert.typeof(vpid, 'string');
+        assert.typeof(rejected, 'boolean');
+        insistCapData(data);
+      }
+      break;
+    }
+    case 'dropExports': {
+      const [slots] = rest;
+      assert(Array.isArray(slots));
+      for (const slot of slots) {
+        assert.typeof(slot, 'string');
+      }
+      break;
+    }
+    default:
+      assert.fail(`unknown delivery type ${type}`);
+  }
+  return vdo;
+}
+
+/**
+ * @param {unknown} vdr
+ * @returns { VatDeliveryResult, asserts vdr is VatDeliveryResult }
```

Comment:
likewise, "An asserts return type predicate implies that the returned value is of type void ..."